### PR TITLE
Ruler querier service option

### PR DESCRIPTION
**Pull request by gonzalez**
#### What this PR does
This PR provides the option to deploy a read path dedicated to the ruler.
This allows you to scale user/dashboards queries independently from the ruler usage.

### CODE REVIEW

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 1]
```diff
@@ -172,6 +172,7 @@ mimir:
     blocks_storage:
       backend: s3
       bucket_store:
+        max_chunk_pool_bytes: 2147483648 # 12GiB
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/values.yaml:175]: Just curious, why is this necessary?

File: operations/helm/charts/mimir-distributed/templates/ruler/ruler-dep.yaml [hunk 2]
```diff
@@ -59,6 +59,10 @@ spec:
             - "-store-gateway.sharding-ring.zone-awareness-enabled=false"
             {{- end }}
           {{- end }}
+          # {{- if .Values.ruler_querier_service.enabled }}
+          #   - "-ruler.remote-evaluation=false"
+          #   - "-ruler.remote-evaluation-migration=false"
+          # {{- end }}
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/templates/ruler/ruler-dep.yaml:65]: Is this left in by mistake?

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 3]
```diff
@@ -1151,8 +1156,151 @@ querier:
   env: []
   extraEnvFrom: []
 
+
+ruler_querier:
+  replicas: 10
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/values.yaml:1161]: 10 replicas seems a bit high for default, considering the user path querier comes with 2. Maybe we default to 2 here as well?

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 4]
```diff
@@ -3517,3 +3739,9 @@ continuous_test:
 #   data: |
 #     extra.yml: "does-my-install-need-extra-info: true"
 extraObjects: []
+
+# -- Add a set of dedicated query-frontend and queriers for the ruler to use.
+# -- This is useful for isolating the ruler queries from other queriers (api/grafana).
+# https://grafana.com/docs/mimir/latest/operators-guide/architecture/components/ruler/#remote
+ruler_querier_service:
+  enabled: false
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/values.yaml:3747]:
```suggestion
remote_rule_evaluation:
  enabled: false
```

I think this name better aligns with the terminology we use elsewhere, WDYT?

Obviously, we'd need to update all the references in this PR as well.

**Review by Logiraptor — COMMENTED** [review]
Overall this looks pretty good. I should have some time tomorrow to give it a try locally. In the meantime I've left a few comments.

File: operations/helm/charts/mimir-distributed/CHANGELOG.md [hunk 5]
```diff
@@ -27,7 +27,7 @@ Entries should be ordered as follows:
 Entries should include a reference to the Pull Request that introduced the change.
 
 ## main / unreleased
-
+* [ENHANCEMENT] Add support for a standalone query path for the ruler. This allows for the isolation of ruler and user query paths. #5081
```
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/CHANGELOG.md:30]: This is more like a quite a feature IMO, not barely an enhancement

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 4 above)
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/values.yaml:3747]: Is it possible to wire this in when the ruler-querier replicas are more than 0?

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 1 above)
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/values.yaml:175]: This option was recently deprecated and removed in #4996 I think it should be removed from this PR too

**Review by dimitarvdimitrov — COMMENTED** [review]
Thanks for the contribution. I reviewed it on a high level and haven't looked at the manifests in detail yet.

I think we should also add docs for using this. We have a page for running the chart in production. This is a good candidate for providing a bit more isolation between rules and the user queries for more sizable clusters. This is the doc I am referring to https://github.com/grafana/mimir/blob/335c64a12f3102fc0399fd9666f7d238ab857168/docs/sources/helm-charts/mimir-distributed/run-production-environment-with-helm/_index.md

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 1 above)
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/values.yaml:175]: @Logiraptor  there's a step in the pipeline that compares jsonnet values to the helm values 
https://github.com/grafana/mimir/actions/runs/5090989185/jobs/9150476748?pr=5081

@dimitarvdimitrov  I will do that thx

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 3 above)
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/values.yaml:1160]: @Logiraptor  agreed thx.
@dimitarvdimitrov It should be off by default.  I can update the small and large values to include the service. 

```
ruler_querier_service:
  enabled: false
```

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 4 above)
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/values.yaml:3746]: With this feature we would have 2 evaluation modes but with 2 implementations of remote_rule_evaluation (shared vs dedicated ruler query paths)   We should probably distinguish those in the name ? Unless we simply have remote evaluation always deploy a dedicated query path ?  @Logiraptor @dimitarvdimitrov

File: operations/helm/charts/mimir-distributed/templates/ruler/ruler-dep.yaml (hunk 2 above)
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/templates/ruler/ruler-dep.yaml]: Yes, thanks.

File: operations/helm/charts/mimir-distributed/templates/ruler-query-frontend/ruler-query-frontend-dep.yaml [hunk 6]
```diff
@@ -0,0 +1,131 @@
+{{- if .Values.ruler_querier_service.enabled -}}
+apiVersion: apps/v1
+kind: Deployment
```
[the remark this answers is not in the document]
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/templates/ruler-query-frontend/ruler-query-frontend-dep.yaml:3]: I don't understand - where are you seeing the duplicate nginx deployments - thanks.

[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/ruler-query-frontend/ruler-query-frontend-dep.yaml:3]: I meant this nginx deployment https://github.com/grafana/mimir/blob/18f1dd062f141dbf21772930380e4fa77fa4bb69/operations/helm/charts/mimir-distributed/templates/nginx/nginx-dep.yaml#L52

and this one https://github.com/grafana/mimir/blob/18f1dd062f141dbf21772930380e4fa77fa4bb69/operations/helm/charts/mimir-distributed/templates/gateway/gateway-dep.yaml#L78

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl [hunk 7]
```diff
@@ -416,6 +416,9 @@ Examples:
   "querier" "querier"
   "query-frontend" "query_frontend"
   "query-scheduler" "query_scheduler"
+  "ruler-querier" "ruler_querier"
+  "ruler-query-frontend" "ruler_query_frontend"
+  "ruler-query-scheduler" "ruler_query_scheduler"
```
Comment (dreamlibrarian) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:421]: Unless I'm very much mistaken, it looks like templates/ruler-query-scheduler  folder is not yet checked in?

Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/templates/_helpers.tpl:421]: You are correct thx - I've pushed the latest update

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 8]
```diff
@@ -334,6 +334,10 @@ mimir:
       alertmanager_url: dnssrvnoa+http://_http-metrics._tcp.{{ template "mimir.fullname" . }}-alertmanager-headless.{{ .Release.Namespace }}.svc.{{ .Values.global.clusterDomain }}/alertmanager
       enable_api: true
       rule_path: /data
+    {{- if .Values.ruler_querier_service.enabled }}
+      query_frontend:
+        address: dns:///{{ template "mimir.fullname" . }}-ruler-query-frontend.{{ .Release.Namespace }}.svc:{{ include "mimir.serverHttpListenPort" .  }}
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/values.yaml:339]:
```suggestion
        address: dns:///{{ template "mimir.fullname" . }}-ruler-query-frontend.{{ .Release.Namespace }}.svc:{{ include "mimir.serverGrpcListenPort" .  }}
```

Tested locally and ruler was throwing a bunch of errors. Traced it back to the fact that should use the grpc port

File: operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      containers:
+        - name: ruler-query-scheduler
+          image: "{{ include "mimir.imageReference" . }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          args:
+            - "-target=query-scheduler"
+            - "-config.expand-env=true"
+            - "-config.file=/etc/mimir/mimir.yaml"
+            - "-server.grpc.keepalive.max-connection-age=2562047h" # 100000 days, effectively infinity
+            - "-server.grpc.keepalive.max-connection-age-grace=2562047h" # 100000 days, effectively infinity
```
Comment (Logiraptor) [inline operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml:48]: What is the purpose of setting such a high value? Seems unsafe since you could leak connections over time.

Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml:48]: @Logiraptor  these are brought over from our query-scheduler templates. What are reasonable values ?

Comment (Logiraptor) [reply operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml:48]: Gotcha, there could be a reason for this. Let me look into it.

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 4 above)
Comment (gonzalez) [reply operations/helm/charts/mimir-distributed/values.yaml:3746]: > Is it possible to wire this in when the ruler-querier replicas are more than 0?

@dimitarvdimitrov  the default values for ruler-querier will always have the replicas > 0

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/values.yaml:3746]: I see. They have > 0 replicas, but are sometimes not created.

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/values.yaml:3746]: But the operator can choose to set the dedicated rulers to 0 replicas and use the shared query path. This will give them e.g. query sharding and query scheduling, which  the ruler itself doesn't support; all without having to deploy more resources.

I suggest to make it possible to deploy in both ways - dedicated and shared. Dedicated when the dedicated query-frontend replicas are more than 0 and shared when they are 0. 

If you agree, then I'd also suggest to make the dedicated ruler path deploy with 0 replicas. The reason is that we aim to have the chart more compatible with smaller deployments, but still enable larger deployments. A dedicated ruler path is a somewhat advanced feature and many regular operators might not need. Once we do that it would be great to document how to setup the remote ruler evaluation both with dedicated query path and a shared query path.

File: operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml (hunk 9 above)
Comment (Logiraptor) [reply operations/helm/charts/mimir-distributed/templates/ruler-query-scheduler/ruler-query-scheduler-dep.yaml:48]: Found my answer here: https://github.com/grafana/mimir/pull/3262

File: operations/helm/charts/mimir-distributed/values.yaml (hunk 4 above)
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/values.yaml:3746]: thoughts on this @gonzalez?

**Comment by lasermoth** [comment]
I'm looking for this functionality currently, it doesn't look like this PR has been touched in a year.
Are we able to progress it?

**Comment by dimitarvdimitrov** [comment]
If @gonzalez doesn't have the bandwidth to take care of this, I think anyone else should be able to complete this building on top of his work.

I haven't looked at the PR in a while, but IIRC the comments I left were the only major blockers to merging this.


It looks like there are a lot of conflicts, but they seem minor to me - the majority of files are just autogenerated manifests; `make build-helm-tests` after rebasing should resolve them

**Comment by alex5517** [comment]
@dimitarvdimitrov, 

I created a new PR which tries to add this feature: #7964
